### PR TITLE
Fix ::class on string expression

### PR DESCRIPTION
**Pull request by rvanvelzen**

### CODE REVIEW

File: src/Rules/Classes/ClassConstantRule.php [hunk 1]
```diff
@@ -130,7 +130,7 @@ public function processNode(Node $node, Scope $scope): array
 					];
 				}
 
-				if ((new StringType())->isSuperTypeOf($classType)->yes()) {
+				if (!$class instanceof Node\Scalar\String_ && (new StringType())->isSuperTypeOf($classType)->yes()) {
```
Comment (staabm) [inline src/Rules/Classes/ClassConstantRule.php]: would this work?

```suggestion
				if (!$class instanceof Node\Scalar\String_ && $classType->isString()->yes()) {
```

Comment (rvanvelzen) [reply src/Rules/Classes/ClassConstantRule.php]: That should definitely work :+1:

**Comment by ondrejmirtes** [comment]
I think this code pattern isn't supported: https://3v4l.org/MM1pJ

**Comment by rvanvelzen** [comment]
Absolutely right, but strangely enough this is: https://3v4l.org/Rb3oU

**Comment by ondrejmirtes** [comment]
Whaat!? :) Sounds like an inconsistency that should be reported in https://github.com/php/php-src/issues :)

I just did: https://github.com/php/doc-en/issues/1646

**Comment by rvanvelzen** [comment]
I'll let this sit here until php/doc-en#1646 is resolved :)

**Comment by ondrejmirtes** [comment]
So the conclusion in https://github.com/php/doc-en/issues/1646 is unequivocal :) Can you please add a test that `'Foo'::class` is not reported by ClassConstantRule and this example is reported?

```php
$foo = 'Foo';
var_dump($foo::class);
```

And maybe also `assertType()` for this example too to make sure we have an ErrorType there.

Thanks :)

**Comment by rvanvelzen** [comment]
The error cases are already covered by tests, so this should be ready to go :)

**Comment by staabm** [comment]
> Absolutely right, but strangely enough this is: [3v4l.org/Rb3oU](https://3v4l.org/Rb3oU)

may I ask how you realized this is a thing? did you see such code somewhere?

**Comment by rvanvelzen** [comment]
I stumbled upon it while fumbling around with phpstan/phpstan#7391 :sweat_smile:

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**